llvm/lib/CAS/BuiltinCAS.h

```diff
@@ -0,0 +1,200 @@
+//===- BuiltinCAS.h ---------------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIB_CAS_BUILTINCAS_H
+#define LLVM_LIB_CAS_BUILTINCAS_H
+
+#include "llvm/ADT/Optional.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/CAS/CASDB.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/SHA1.h"
+#include <cstddef>
+
+namespace llvm {
+namespace cas {
+namespace builtin {
+
+/// Dereference the inner value in \p E, adding an Optional to the outside.
+/// Useful for stripping an inner Optional in return chaining.
+///
+/// \code
+/// Expected<Optional<SomeType>> f1(...);
+///
+/// Expected<SomeType> f2(...) {
+///   if (Optional<Expected<NoneType>> E = dereferenceValue(f1()))
+///     return std::move(*E);
+///
+///   // Deal with None...
+/// }
+/// \endcode
+template <class T>
+inline Optional<Expected<T>> dereferenceValue(Expected<Optional<T>> E) {
+  if (!E)
+    return Expected<T>(E.takeError());
+  if (*E)
+    return Expected<T>(std::move(**E));
+  return None;
+}
+
+/// Dereference the inner value in \p E, generating an error on failure.
+///
+/// \code
+/// Expected<Optional<SomeType>> f1(...);
+///
+/// Expected<SomeType> f2(...) {
+///   if (Optional<Expected<NoneType>> E = dereferenceValue(f1()))
+///     return std::move(*E);
+///
+///   // Deal with None...
+/// }
+/// \endcode
+template <class T>
+inline Expected<T> dereferenceValue(Expected<Optional<T>> E,
+                                    function_ref<Error()> OnNone) {
+  if (Optional<Expected<T>> MaybeExpected = dereferenceValue(std::move(E)))
+    return std::move(*MaybeExpected);
+  return OnNone();
+}
+
+/// If \p E and \c *E, move \c **E into \p Sink.
+///
+/// Enables expected and optional chaining in one statement:
+///
+/// \code
+/// Expected<Optional<Type1>> f1(...);
+///
+/// Expected<Optional<Type2>> f2(...) {
+///   SomeType V;
+///   if (Optional<Expected<NoneType>> E = moveValueInto(f1(), V))
+///     return std::move(*E);
+///
+///   // Deal with value...
+/// }
+/// \endcode
+template <class T, class SinkT>
+inline Optional<Expected<NoneType>>
+moveValueInto(Expected<Optional<T>> ExpectedOptional, SinkT &Sink) {
+  if (!ExpectedOptional)
+    return Expected<NoneType>(ExpectedOptional.takeError());
+  if (!*ExpectedOptional)
+    return Expected<NoneType>(None);
+  Sink = std::move(*ExpectedOptional);
+  return None;
+}
+
+inline StringRef toStringRef(ArrayRef<char> Data) {
+  return StringRef(Data.data(), Data.size());
+}
+
+inline ArrayRef<char> toArrayRef(StringRef Data) {
```

Comment:
Similarly there is an `arrayRefFromStringRef` in ADT.  This should probably have the same name.

Reply:
Best part about moving code is getting it (finally) reviewed! Thanks, I'll get this stuff cleaned up.

Reply:
The same name would be super awkward since it's the same parameters... makes me wonder if we should just use `ArrayRef<uint8_t>` to hack around it... since StringExtras.h is optimized for that case.